lib/Dialect/Moore/Transforms/MergeAssignments.cpp

```diff
@@ -0,0 +1,79 @@
+//===- MergeAssignments.cpp - Merge declaration and assignment ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the MergeAssignments pass.
+// Find easy uses of declaration and assignments and merge them into
+// assigned_variable. Easy use represents the declaration doesn't be performed
+// bit slice. Like bit [127:0] b; assign b = 128'b0;
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Moore/MooreOps.h"
+#include "circt/Dialect/Moore/MoorePasses.h"
+#include "mlir/IR/Dialect.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_MERGEASSIGNMENTS
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+
+namespace {
+struct MergeAssignmentsPass
+    : public circt::moore::impl::MergeAssignmentsBase<MergeAssignmentsPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+std::unique_ptr<mlir::Pass> circt::moore::createMergeAssignmentsPass() {
+  return std::make_unique<MergeAssignmentsPass>();
+}
+
+// TODO: The net can be driven multiple times. However, the related rule is
+// complicated. So we will implement this in the future.
+
+// Only collect the easy declaration and its value at module level.
+static void collectAssignmets(SVModuleOp moduleOp,
+                              DenseMap<Value, Value> &assignments) {
+  moduleOp->walk([&](Operation *op) {
```

Comment:
You don't need inner TypeSwitch. 

```suggestion
  moduleOp->walk([&](ContinuousAssignOp op) {
```

Reply:
From the perspective of the design of `AssignedVarOp`, I think we can only handle continuous assignments. And due to the dominance of non-continuous assignments. We can not merge variable/net and continuous assignment into one op called `assigned_var`. I'll tweak it later.

Reply:
> From the perspective of the design of AssignedVarOp, I think we can only handle continuous assignments.

I meant ` moduleOp->walk([&](ContinuousAssignOp op) {` is equivalent to `moduleOp->walk([&](Operation*op) {
TypeSwith(op)...`.

Reply:
I understood. I added TypeSwitch here to reserve a case branch to handle some situations I have not considered now. Because I'm not sure whether I think enough. But then I found that non-continuous assignments would trigger the dominance error. So I left this comment.